sbin/slattach/slattach.8
.\"	BSDI $Id: slattach.8,v 1.3 1993/12/19 19:26:36 sanders Exp $
.\"
.\" Copyright (c) 1986, 1991 The Regents of the University of California.
.\" All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\" 3. All advertising materials mentioning features or use of this software
.\"    must display the following acknowledgement:
.\"	This product includes software developed by the University of
.\"	California, Berkeley and its contributors.
.\" 4. Neither the name of the University nor the names of its contributors
.\"    may be used to endorse or promote products derived from this software
.\"    without specific prior written permission.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
.\" ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
.\" FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.\"     @(#)slattach.8	6.4 (Berkeley) 3/16/91
.\"
.Dd March 16, 1991
.Dt SLATTACH 8
.Os BSD 4.3
.Sh NAME
.Nm slattach
.Nd attach serial lines as network interfaces
.Sh SYOPNSIS
.Nm slattach
.Ar ttyname Op Ar baudrate
.Sh DESCRIPTION
.Nm slattach
is used to assign a tty line to a network interface,
and to define the network source and destination addresses.
The following operands are supported by
.Nm slattach :
.Bl -tag -width Ar
.It Ar ttyname
Specifies the name of the tty device.
.Ar ttyname
should be a string of the form
.Ql ttyXX ,
or
.Ql /dev/ttyXX .
.It Ar baudrate
Specifies the speed of the connection.  If not specified, the
default of 9600 is used.
.El
.Pp
Only the super-user may attach a network interface.
.Pp
To detach the interface, use
.Dq Li ifconfig interface-name down
after killing off the
.Nm slattach
process.
.Ar interface-name
is the name that is shown by
.Xr netstat 1 .
.Pp
.Nm Slattach
places the specified tty in raw mode using
.Xr cfmakeraw 3 .
.Pp
.Nm Slattach
is intended to be used to run SLIP over dedicated lines (typically used in
.Pa /etc/netstart ) .
For dial-up services use
.Xr startslip 1 .
.Sh EXAMPLE
.Bd -literal -offset indent -compact
# Dedicated 57600-baud SLIP on /dev/tty01
ifconfig sl0 myhost rmthost link2 up
stty -f /dev/tty01 cts_oflow rts_iflow
slattach /dev/tty01 57600
.Ed
.Sh DIAGNOSTICS
Messages indicating the specified interface does not exit, the
requested address is unknown, the user is not privileged and
tried to alter an interface's configuration.
.Sh SEE ALSO
.Xr startslip 1 ,
.Xr netstat 1 ,
.Xr netintro 4 ,
.Xr ifconfig 8 ,
.Xr rc 8 ,
.Xr sliplogin 8
.Sh HISTORY
The
.Nm
command appeared in
.Bx 4.3 .


sbin/slattach/slattach.c
/*-
 * Copyright (c) 1993 Berkeley Software Design, Inc. All rights reserved.
 * The Berkeley Software Design Inc. software License Agreement specifies
 * the terms and conditions for redistribution.
 *
 *	BSDI $Id: slattach.c,v 1.3 1993/12/19 19:26:37 sanders Exp $
 */
 
/*
 * Copyright (c) 1988 Regents of the University of California.
 * All rights reserved.
 *
 * This code is derived from software contributed to Berkeley by
 * Rick Adams.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 3. All advertising materials mentioning features or use of this software
 *    must display the following acknowledgement:
 *	This product includes software developed by the University of
 *	California, Berkeley and its contributors.
 * 4. Neither the name of the University nor the names of its contributors
 *    may be used to endorse or promote products derived from this software
 *    without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 */

#ifndef lint
char copyright[] =
"@(#) Copyright (c) 1988 Regents of the University of California.\n\
 All rights reserved.\n";
#endif /* not lint */

#ifndef lint
static char sccsid[] = "@(#)slattach.c	4.6 (Berkeley) 6/1/90";
#endif /* not lint */

#include <sys/param.h>
#include <sys/ioctl.h>
#include <sys/fcntl.h>
#include <stdio.h>
#include <string.h>
#include <termios.h>
#include <paths.h>
#include <errno.h>
#include <err.h>

#define DEFAULT_BAUD	9600

int slipdisc = SLIPDISC;

char devname[MAXPATHLEN];
char hostname[MAXHOSTNAMELEN];

main(argc, argv)
	int argc;
	char *argv[];
{
	register int fd;
	register char *dev = argv[1];
	struct termios tio;
	speed_t speed;

	if (argc < 2 || argc > 3)
		errx(1, "Usage: %s ttyname [baudrate]\n", argv[0]);
	speed = argc == 3 ? atoi(argv[2]) : DEFAULT_BAUD;

	if (strncmp(_PATH_DEV, dev, sizeof(_PATH_DEV) - 1)) {
		(void) sprintf(devname, "%s/%s", _PATH_DEV, dev);
		dev = devname;
	}
	/* force the open with O_NONBLOCK and then set CLOCAL later */
	if ((fd = open(dev, O_RDWR | O_NONBLOCK)) < 0)
		err(1, "%s", dev);
	if (tcgetattr(fd, &tio) != 0)
		err(1, "%s", dev);
	cfsetspeed(&tio, speed);
	cfmakeraw(&tio);
	tio.c_cflag |= CLOCAL;
	if (tcsetattr(fd, TCSAFLUSH, &tio) != 0)
		err(1, "%s", dev);
	if (fcntl(fd, F_SETFL, 0) == -1)	/* clear O_NONBLOCK */
		err(1, "%s", dev);
	if (ioctl(fd, TIOCSETD, &slipdisc) < 0)
		err(1, "%s", dev);
	daemon(0, 1);
	while (1)
		sigpause(0L);
}
